dali/pipeline/operator/builtin/split.cc

```diff
@@ -0,0 +1,109 @@
+      make_string("Split description must cover whole input, got ", input.num_samples(),
+                  " input samples and ", predicate.num_samples(), " elements denoting the split."));
+  DALI_ENFORCE(predicate.shape().sample_dim() == 0, "Only scalar indexing is supported.");
+
+  category_counts_.fill(0);
+
+  for (int i = 0; i < predicate.num_samples(); i++) {
+    int output_category_idx = get_category_index(predicate, i);
+    category_counts_[output_category_idx]++;
+  }
+
+  // TODO(klecki): we can construct the output_desc, it won't be useful now
+  return false;
+}
+
+
+template <typename Backend>
+void Split<Backend>::RunImpl(Workspace &ws) {
+  const auto &input = ws.template Input<Backend>(0);
+  const auto &predicate = ws.ArgumentInput("predicate");
+  auto sample_idx_in_output = uniform_array<kMaxCategories>(0);
+
+  for (int output_category_idx = 0; output_category_idx < kMaxCategories; output_category_idx++) {
+    auto &output = ws.template Output<Backend>(output_category_idx);
+
+    // We can (and need to) do it only once, for each new output instance, when it doesn't have
+    // data yet. It should be consistent across iterations.
+    if (!output.has_data()) {
+      output.set_type(input.type());
+      output.set_sample_dim(input.shape().sample_dim());
+      output.SetLayout(input.GetLayout());
+      output.set_device_id(input.device_id());
+      output.set_pinned(input.is_pinned());
+      // We let the executor set the desired order, the rest is propagated from the input.
+    }
+    output.SetSize(category_counts_[output_category_idx]);
+  }
+
+  for (int input_sample_idx = 0; input_sample_idx < predicate.num_samples(); input_sample_idx++) {
+    int output_category_idx = get_category_index(predicate, input_sample_idx);
+    auto &output = ws.template Output<Backend>(output_category_idx);
+
+    // get the output index and increment for the next sample.
+    int output_sample_idx = sample_idx_in_output[output_category_idx];
+    sample_idx_in_output[output_category_idx]++;
+
+    // share the sample to the output
+    output.SetSample(output_sample_idx, input, input_sample_idx);
+  }
+  WriteTestsDiagnostics(ws);
+}
+
+
+template <typename Backend>
+void Split<Backend>::RegisterTestsDiagnostics() {
+  this->RegisterDiagnostic("input_pinned", &in_pinned_);
+  this->RegisterDiagnostic("output_0_pinned", &out_0_pinned_);
+  this->RegisterDiagnostic("output_1_pinned", &out_1_pinned_);
+}
+
+template <typename Backend>
+void Split<Backend>::WriteTestsDiagnostics(const Workspace &ws) {
+  in_pinned_ = ws.template Input<Backend>(0).is_pinned();
+  out_0_pinned_ = ws.template Output<Backend>(0).is_pinned();
+  out_1_pinned_ = ws.template Output<Backend>(1).is_pinned();
+}
+
+DALI_SCHEMA(experimental___Split)
+    .DocStr(R"code(Split batch based on a predicate.)code")
+    .NumInput(1)
+    .NumOutput(2)
+    .AddArg("predicate", "Boolean categorization of the input batch", DALI_BOOL, true)
```

Comment:
I'd consider something like `partition` `group` (whatever we choose that works for both Split and Merge) and make it an index rather than a bool

Reply:
I think that what you suggest is planned as an extension of this functionality. Since it makes it just a little bit easier I'd be in favor of proceeding with the boolean predicate for now.

Reply:
As discussed, we keep the boolean predicate for now. 
The integer `group` would be one of the next steps in this effort. Accepting booleans for conditions is needed, `group` would be a mutually exclusive argument with it, so we can easily extend. But handling arbitrary number of groups adds a lot of complexity on DALI side and probably requires more API extensions(to handle variadic number of outputs we probably need additional argument), and complicates testing.